### PR TITLE
feat: support training with FSDP + SGLang/vLLM, support VLM datasets and models, add reward worker

### PR DESCRIPTION
**Pull request by guozhen1997**
### Description

1. For reasoning tasks, support training with FSDP + SGLang/vLLM
2. Support VLM datasets and models
3. Add vqa reward function for VLM RL training
4. Add reward worker to support both rule-based reward function and reward model

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: .github/workflows/code-test.yml [hunk 1]
```diff
@@ -195,6 +224,26 @@ jobs:
...
+  qwen-vl-grpo-test:
+    needs: [check-changes]
+    if: needs.check-changes.outputs.file_filter == 'true'
+    runs-on: reason
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v5
+
+      - name: FSDP SGLang Collocated mode
+        run: |
```
Comment (andylin-hao) [inline .github/workflows/code-test.yml:238]: Need to add timeouts

File: rlinf/algorithms/rewards/vqa/__init__.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            qa_accuracy_reward,
+            think_format_reward,
+            answer_format_reward,
+        ]
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def get_reward(self, completions: List[str], answers: List[dict]) -> List[float]:
+        rewards = []
+        for i, reward_function in enumerate(self.reward_functions):
```
Comment (Lin-xs) [inline rlinf/algorithms/rewards/vqa/__init__.py]: Use `zip` to replace enumerate.

File: rlinf/algorithms/rewards/__init__.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def get_reward_class(name: str):
+    assert name in reward_registry, f"Reward {name} not found"
+    return reward_registry[name]
+
+
+reward_registry = {}
+
+register_reward("math", MathReward)
```
Comment (andylin-hao) [inline rlinf/algorithms/rewards/__init__.py:32]: Check algorithm/registry.py. Remove duplicate register_reward

**Comment by andylin-hao** [comment]
Need to simplify the rollout result splitting. Split the data input instead of the result, as input data is much simpler than rollout result.

File: rlinf/algorithms/rewards/vqa/__init__.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import torch
+from omegaconf import DictConfig
+
+from .format_rewards import answer_format_reward, think_format_reward
+from .qa_rewards import qa_accuracy_reward
+
+
+class VQAReward:
+    def __init__(self, config: DictConfig):
+        reward_weights_config = config.get(
```
Comment (Lin-xs) [inline rlinf/algorithms/rewards/vqa/__init__.py]: Better to check whether these three fields are set in config.

**Comment by andylin-hao** [comment]
Need to check Megatron's TP being larger than Rollout.

File: rlinf/algorithms/rewards/__init__.py (hunk 3 above)
Comment (anHappyDog) [reply rlinf/algorithms/rewards/__init__.py:32]: done

File: rlinf/algorithms/rewards/vqa/__init__.py (hunk 2 above)
Comment (anHappyDog) [reply rlinf/algorithms/rewards/vqa/__init__.py]: done

File: rlinf/algorithms/rewards/vqa/__init__.py (hunk 4 above)
Comment (anHappyDog) [reply rlinf/algorithms/rewards/vqa/__init__.py]: done

File: rlinf/hybrid_engines/vllm/vllm_0_8_5/worker.py [hunk 5]
```diff
@@ -72,6 +75,14 @@ def initialize_from_config(self, kv_cache_config: KVCacheConfig) -> None:
         super().initialize_from_config(kv_cache_config)
 
     def offload_model_weights(self) -> None:
+        torch.cuda.synchronize()
```
Comment (Lin-xs) [inline rlinf/hybrid_engines/vllm/vllm_0_8_5/worker.py]: delete

File: rlinf/workers/actor/megatron_actor_worker.py [hunk 6]
```diff
@@ -1033,7 +963,10 @@ def sync_model_to_rollout(self):
     def _compute_rollout_metrics(self, batch):
         rollout_metrics, total_prompt_lengths, total_decode_lengths = (
             compute_rollout_metrics(
-                batch, self.cfg.data.max_prompt_length, self.response_len
+                batch,
+                self.cfg.data.max_prompt_length,
+                self.response_len,
+                self._world_size,
```
Comment (Lin-xs) [inline rlinf/workers/actor/megatron_actor_worker.py:969]: Now this function will call `torch.dist.all_gather` across all ranks in the world, rather than `data_parallel_group`, is this correct?

File: rlinf/hybrid_engines/vllm/vllm_0_8_5/worker.py (hunk 5 above)
Comment (anHappyDog) [reply rlinf/hybrid_engines/vllm/vllm_0_8_5/worker.py]: done

File: rlinf/workers/actor/megatron_actor_worker.py (hunk 6 above)
Comment (anHappyDog) [reply rlinf/workers/actor/megatron_actor_worker.py:969]: done

**Comment by anHappyDog** [comment]
image
